src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -953,25 +1064,42 @@ namespace cryptonote
     else
       stem_txs.reserve(arg.txs.size());
 
-    for (auto& tx : arg.txs)
+    for (auto& tx_blob : arg.txs)
     {
       tx_verification_context tvc{};
-      if (!m_core.handle_incoming_tx(tx, tvc, tx_relay, true) && !tvc.m_no_drop_offense)
+      if (!m_core.handle_incoming_tx(tx_blob, tvc, tx_relay, true))
       {
-        LOG_PRINT_CCONTEXT_L1("Tx verification failed, dropping connection");
-        drop_connection(context, false, false);
-        return 1;
+        if (!tvc.m_no_drop_offense)
+        {
+          LOG_PRINT_CCONTEXT_L1("Tx verification failed, dropping connection");
+          drop_connection(context, false, false);
+          return 1;
+        }
+      }
+      else
+      {
+        context.add_received();
+        // This is extremely inefficient. We have to change m_core.handle_incoming_tx
+        // we are already parsing inside handle_incoming_tx
+        transaction tx;
+        crypto::hash tx_hash;
+        if (cryptonote::parse_and_validate_tx_from_blob(tx_blob, tx, tx_hash)
+            && m_request_manager.already_requested_tx(tx_hash))
+        {
+          m_request_manager.received_requested(context.m_connection_id, tx_hash);
+          context.remove_in_flight_request();
```

Comment:
You don't check the return value of `received_requested` here. If it returns `false`, then tx_hash wasn't received and you shouldn't remove in flight request, no?